### PR TITLE
adding a blog about statefulset

### PR DESCRIPTION
**Pull request by linsun**
[ ] Configuration Infrastructure
[ x ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

### CODE REVIEW

**Comment by linsun** [comment]
We have done a few technical reviews from co-authors of this blog (@christian-posta @howardjohn @hzxuzhonghu), it would be ready for doc review as soon as I get lint/etc to pass.

File: content/en/blog/2021/stateful-set-made-easier/index.md
Comment (craigbox) [inline content/en/blog/2021/stateful-set-made-easier/index.md]:
```suggestion
Are my ZooKeeper services working and is the status `READY`? Let’s find out! ZooKeeper listens on 3 ports:

*   port 2181 is the TCP port for clients to connect to the ZooKeeper service
```

doesn't render the list properly

File: content/en/blog/2021/stateful-set-made-easier/index.md
Comment (craigbox) [inline content/en/blog/2021/stateful-set-made-easier/index.md:118]: you say Istio 1.10 a lot in this Istio 1.10 sentence, powered by Istio 1.10

File: content/en/blog/2021/stateful-set-made-easier/index.md
Comment (craigbox) [inline content/en/blog/2021/stateful-set-made-easier/index.md:122]: Would it be more powerful if you installed again from the top?

(In fact, if that's the case, maybe you should lead with this section?)

File: content/en/blog/2021/stateful-set-made-easier/index.md
Comment (linsun) [reply content/en/blog/2021/stateful-set-made-easier/index.md:19]: Actually, it is not that simple. :) I think ZK doesn't recommend users to enable this in production.  This quorumListenOnAllIPs is an unsafe option per their doc, even tho our istio [doc](https://istio.io/latest/faq/applications/#zookeeper) recommended users to do it today in 1.9, which is bad.

File: content/en/blog/2021/stateful-set-made-easier/index.md
Comment (linsun) [reply content/en/blog/2021/stateful-set-made-easier/index.md:15]: This is true for stateful sets in general.  also mentioned in @howardjohn 's blog. I think it is related to how stateful sets pods are created from the same spec and didn't want to listen on loopback interface for that inter-pod communication.

Comment (howardjohn) [reply content/en/blog/2021/stateful-set-made-easier/index.md:15]: This is not accurate, any application, statefulset or otherwise, can bind to any interface it chooses. It just happens to be common among applications that happen to use statefulsets

File: content/en/blog/2021/stateful-set-made-easier/index.md
Comment (linsun) [reply content/en/blog/2021/stateful-set-made-easier/index.md]: updated

File: content/en/blog/2021/stateful-set-made-easier/index.md
Comment (linsun) [reply content/en/blog/2021/stateful-set-made-easier/index.md]: updated

File: content/en/blog/2021/stateful-set-made-easier/index.md
Comment (linsun) [reply content/en/blog/2021/stateful-set-made-easier/index.md]: Yep good catch!

File: content/en/blog/2021/stateful-set-made-easier/index.md
Comment (linsun) [reply content/en/blog/2021/stateful-set-made-easier/index.md:121]: Open to suggestions on changing this :) I used 1.10 few times for clarity.

File: content/en/blog/2021/stateful-set-made-easier/index.md
Comment (linsun) [reply content/en/blog/2021/stateful-set-made-easier/index.md:125]: My thoughts is to meet users where they are today given these are very common issues with stateful sets and proved that we are better while explaining how the network behavior changed.

File: content/en/blog/2021/stateful-set-made-easier/index.md
Comment (linsun) [reply content/en/blog/2021/stateful-set-made-easier/index.md]: @craigbox @howardjohn anything you propose to change here?  I didn't cover all pods/apps as John covered in his other blog.  I think what we have around stateful sets is accurate in this blog.

Comment (howardjohn) [reply content/en/blog/2021/stateful-set-made-easier/index.md]: `Pods within a StatefulSet listen on the pod IP only instead of `0.0.0.0` for communication within the StatefulSet pods.` is not accurate. It would be accurate to say `It is common for pods within a StatefulSet listen on the pod IP only, instead of 0.0.0.0, for communication within the StatefulSet pods`, but it certainly doesn't apply to all statefulsets.

Statefulsets are actually just normal pods with a few fields set BTW, you can do everything with a statefulset with Pod

Comment (linsun) [reply content/en/blog/2021/stateful-set-made-easier/index.md]: Yah, I like that, thanks for that clarification! you'll never know how users can configure a StatefulSet :)

File: .spelling
Comment (ericvn) [inline .spelling:260]: In other places, we typically use backticks around DestinationRule.

File: .spelling
Comment (ericvn) [inline .spelling:704]: We also use backticks around StatefulSet.

File: content/en/blog/2021/stateful-set-made-easier/index.md
Comment (ericvn) [inline content/en/blog/2021/stateful-set-made-easier/index.md:9]: I will also add that Pod usually has backticks as well, at least when capitalized.

File: content/en/blog/2021/stateful-set-made-easier/index.md
Comment (ericvn) [inline content/en/blog/2021/stateful-set-made-easier/index.md:11]: Backtick on ServiceEntry

File: content/en/blog/2021/stateful-set-made-easier/index.md
Comment (ericvn) [inline content/en/blog/2021/stateful-set-made-easier/index.md:13]: **to** listen

**Review by ericvn — COMMENTED** [review]
Some nits on using backticks

**Comment by linsun** [comment]
> Some nits on using backticks

Thanks Eric, all resolved, I think!

File: content/en/blog/2021/stateful-set-made-easier/index.md
Comment (craigbox) [inline content/en/blog/2021/stateful-set-made-easier/index.md:2]: Could you rename the file "statefulsets-made-easier" to match the CamelCase?

File: content/en/blog/2021/stateful-set-made-easier/index.md
Comment (craigbox) [inline content/en/blog/2021/stateful-set-made-easier/index.md:21]:
```suggestion
What is unique about using a `StatefulSet` with a service mesh? The `StatefulSet` pods are created from the same spec, but are not interchangeable: each has a persistent identifier that it maintains across any rescheduling. The kind of apps that run in a `StatefulSet` are often those that need to communicate amongst their pods, and, as they come from a world of hard-coded IP addresses, may listen on the pod IP only, instead of `0.0.0.0`.

ZooKeeper, for example, is configured by default not to listen on all IPs for quorum communication:

{{< text plain >}}
quorumListenOnAllIPs=false
{{< /text >}}

Over the last few releases, the Istio community has [reported many issues](https://github.com/istio/istio/issues/10659) around support for applications running in `StatefulSets`.
```

Thoughts on that? @linsun @howardjohn

File: content/en/blog/2021/stateful-set-made-easier/index.md
Comment (craigbox) [inline content/en/blog/2021/stateful-set-made-easier/index.md:25]:
```suggestion
## `StatefulSets` in action, prior to Istio 1.10

In a GKE cluster running Kubernetes 1.19, we have Istio 1.9.5 installed. We enabled automatic sidecar injection in the `default` namespace, then we installed ZooKeeper using the [Helm charts provided by Bitnami](https://artifacthub.io/packages/helm/bitnami/zookeeper), along with our `sleep` sample application:
```

File: content/en/blog/2021/stateful-set-made-easier/index.md
Comment (craigbox) [inline content/en/blog/2021/stateful-set-made-easier/index.md:55]:
```suggestion
By default, the ZooKeeper installation configures port 2181 to listen on `0.0.0.0` but ports 2888 and 3888 only listen on the pod IP. Let’s check out the network status on each of these ports from one of the ZooKeeper pods:
```

File: content/en/blog/2021/stateful-set-made-easier/index.md
Comment (craigbox) [inline content/en/blog/2021/stateful-set-made-easier/index.md:117]:
```suggestion
What is interesting here is that the inbound on port 3888 has `127.0.0.1` as its endpoint. This is because the Envoy proxy, in versions of Istio prior to 1.10, redirects the inbound traffic to the `loopback` interface, as described in the [our blog post about the change](/blog/2021/upcoming-networking-changes/).
```

File: content/en/blog/2021/stateful-set-made-easier/index.md
Comment (craigbox) [inline content/en/blog/2021/stateful-set-made-easier/index.md]:
```suggestion
## StatefulSets in action with Istio 1.10
```

Comment (craigbox) [reply content/en/blog/2021/stateful-set-made-easier/index.md]: You might need the backticks - I want it on the record that I think it's ugly as sin that we have to use them

File: content/en/blog/2021/stateful-set-made-easier/index.md
Comment (craigbox) [inline content/en/blog/2021/stateful-set-made-easier/index.md:121]:
```suggestion
Now, we have upgraded our cluster to Istio 1.10 and configured the `default` namespace to enable 1.10 sidecar injection. Let’s rolling restart the ZooKeeper `StatefulSet` to update the pods to use the new version of the sidecar proxy:
```

(Perhaps we should say how we did this?  I'm assuming from the context this is an upgrade, not a second cluster)

File: content/en/blog/2021/stateful-set-made-easier/index.md
Comment (craigbox) [inline content/en/blog/2021/stateful-set-made-easier/index.md:127]:
```suggestion
Once the ZooKeeper pods reach the `Running` status, we can check out the network connections for our 3 ports from any of the ZooKeeper pods:
```

File: content/en/blog/2021/stateful-set-made-easier/index.md
Comment (craigbox) [inline content/en/blog/2021/stateful-set-made-easier/index.md:156]:
```suggestion
There are `ESTABLISHED` connections on both port 2888 and 3888!  Next, let us check out the ZooKeeper server status:
```

File: content/en/blog/2021/stateful-set-made-easier/index.md
Comment (craigbox) [inline content/en/blog/2021/stateful-set-made-easier/index.md:167]:
```suggestion
The ZooKeeper service is now running!

We can connect to each of the ZooKeeper pods from the `sleep` pod and run the below command to discover the server status of each pod within the `StatefulSet`. Note that there is no need to create ServiceEntry resources for any of the ZooKeeper pods and we can call these pods directly using their DNS names (e.g. `my-release-zookeeper-0.my-release-zookeeper-headless`) from the `sleep` pod.
```

File: content/en/blog/2021/stateful-set-made-easier/index.md
Comment (craigbox) [inline content/en/blog/2021/stateful-set-made-easier/index.md:206]:
```suggestion
Now our ZooKeeper service is running, let’s use Istio to secure all communication to our regular and headless services. Apply mutual TLS to the `default` namespace:
```

File: content/en/blog/2021/stateful-set-made-easier/index.md
Comment (craigbox) [inline content/en/blog/2021/stateful-set-made-easier/index.md:222]: describe what you're seeing here?

"The padlock icons on the traffic flows indicate that the connections are secure."

File: content/en/blog/2021/stateful-set-made-easier/index.md
Comment (ericvn) [inline content/en/blog/2021/stateful-set-made-easier/index.md:17]: nit: s/not to/to not/

File: content/en/blog/2021/stateful-set-made-easier/index.md
Comment (craigbox) [inline content/en/blog/2021/stateful-set-made-easier/index.md:226]:
```suggestion
With the new networking changes in Istio 1.10, a Kubernetes pod with a sidecar has the same networking behavior as a pod without a sidecar. This change enables stateful applications to function properly in Istio. as we have shown you in this post. We believe this is a huge step towards Istio’s goal of providing transparent service mesh and zero-configuration Istio.
```

**Review by craigbox — COMMENTED** [review]
much better; and hopefully this will make it much better-er

File: content/en/blog/2021/stateful-set-made-easier/index.md
Comment (ericvn) [inline content/en/blog/2021/stateful-set-made-easier/index.md]: nit: along with Istio's `sleep`

File: content/en/blog/2021/stateful-set-made-easier/index.md
Comment (linsun) [reply content/en/blog/2021/stateful-set-made-easier/index.md:21]: Only thing I'm not sure is `and, as they come from a world of hard-coded IP addresses,` otherwise LGTM

@howardjohn ?

**Review by ericvn — COMMENTED** [review]
Some minor nits.

File: content/en/blog/2021/stateful-set-made-easier/index.md
Comment (howardjohn) [reply content/en/blog/2021/stateful-set-made-easier/index.md:21]: Sounds good enough to me, but I am not familiar with non-k8s world :slightly_smiling_face:

File: content/en/blog/2021/stateful-set-made-easier/index.md
Comment (linsun) [reply content/en/blog/2021/stateful-set-made-easier/index.md]: I think okay to using `our` given the authors represent istio community here? @craigbox thoughts?

File: content/en/blog/2021/stateful-set-made-easier/index.md
Comment (linsun) [reply content/en/blog/2021/stateful-set-made-easier/index.md:21]: Thanks Craig, we'll take your suggestions!

File: content/en/blog/2021/stateful-set-made-easier/index.md
Comment (linsun) [reply content/en/blog/2021/stateful-set-made-easier/index.md]: defer to @craigbox since he just changed this portion.

File: content/en/blog/2021/stateful-set-made-easier/index.md
Comment (linsun) [reply content/en/blog/2021/stateful-set-made-easier/index.md]: Thanks, added.

File: content/en/blog/2021/stateful-set-made-easier/index.md
Comment (linsun) [reply content/en/blog/2021/stateful-set-made-easier/index.md]: done!

Comment (linsun) [reply content/en/blog/2021/stateful-set-made-easier/index.md]: Actually, @craigbox  this will cause lint error, so I'm rolling back the change.  This is probably why all the other blog folders use lower case only.

```
http://localhost:5153/blog/2021/statefulsets-made-easier/
45
  [404] http://localhost:5153/blog/2021/StatefulSets-made-easier/view-zookeeper-from-kiali.png
46
ERROR: Detected 1 broken links. Scanned 4000 links in 48.742 seconds.
```

Comment (craigbox) [reply content/en/blog/2021/stateful-set-made-easier/index.md]: Lower case please - "statefulsets-made-easier" should work.  (I was asking to remove the hyphen, not change the case 🙂)

File: content/en/blog/2021/stateful-set-made-easier/index.md
Comment (craigbox) [reply content/en/blog/2021/stateful-set-made-easier/index.md]: note: someone should update the FAQ

File: content/en/blog/2021/statefulsets-made-easier/index.md
Comment (craigbox) [inline content/en/blog/2021/statefulsets-made-easier/index.md:15]:
```suggestion
ZooKeeper, for example, is configured by default to not listen on all IPs for quorum communication:
```

File: content/en/blog/2021/statefulsets-made-easier/index.md
Comment (craigbox) [inline content/en/blog/2021/statefulsets-made-easier/index.md:121]:
```suggestion
## `StatefulSets` in action with Istio 1.10

Now, we have upgraded our cluster to Istio 1.10 and configured the `default` namespace to enable 1.10 sidecar injection. Let’s rolling restart the ZooKeeper `StatefulSet` to update the pods to use the new version of the sidecar proxy:
```

File: content/en/blog/2021/stateful-set-made-easier/index.md
Comment (craigbox) [reply content/en/blog/2021/stateful-set-made-easier/index.md]: I agree with Eric, adding suggestion

File: content/en/blog/2021/stateful-set-made-easier/index.md
Comment (craigbox) [reply content/en/blog/2021/stateful-set-made-easier/index.md]: I'll clarify in a suggestion

File: content/en/blog/2021/stateful-set-made-easier/index.md
Comment (linsun) [reply content/en/blog/2021/stateful-set-made-easier/index.md]: Yes, plan to update it after this is merged. :)

**Review by ericvn — APPROVED** [review]
LGTM - Thanks Lin.

**Review by craigbox — APPROVED** [review]

**Pull request merged**
